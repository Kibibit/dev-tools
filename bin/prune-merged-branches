#!/usr/bin/env node
const { pruneMergedBranches } = require('../lib/prune-merged-branches');
pruneMergedBranches();