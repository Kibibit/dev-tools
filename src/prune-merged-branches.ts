import Table from 'cli-table';
import inquirer from 'inquirer';
import { chain,trim } from 'lodash';
import { BranchSummaryBranch } from 'simple-git';
import simpleGit, { BranchSummary } from 'simple-git/promise';

import { MAIN_BRANCHES } from './main-branch';

interface IExtraData { isMerged?: boolean; remoteName?: string }

type BranchWithExtras = BranchSummaryBranch & IExtraData;

const git = simpleGit();
const remoteInfoRegex = /^\[(.*?)\]\s/g;

const chars = {
  'top': '═', 'top-mid': '╤', 'top-left': '╔', 'top-right': '╗'
  , 'bottom': '═', 'bottom-mid': '╧', 'bottom-left': '╚', 'bottom-right': '╝'
  , 'left': '║', 'left-mid': '╟', 'mid': '─', 'mid-mid': '┼'
  , 'right': '║', 'right-mid': '╢', 'middle': '│'
};

export async function pruneMergedBranches() {
  try {
    await git.fetch(['-p']); // prune? is it necassery?
    const branchSummaryResult = await git.branch(['-vv']);
    const localBranches = branchSummaryResult.branches;
    const localBranchesWithGoneRemotes: BranchWithExtras[] =
      chain(localBranches)
        .filter((item) => !MAIN_BRANCHES.includes(item.name))
        .forEach((item: BranchWithExtras) => {
          // console.log('an item appeared!', item);
          const remoteInfo = item.label.match(remoteInfoRegex);

          if (remoteInfo) {
            const parsedRemoteInfo = trim(remoteInfo[0], '[] ');
            const isMerged = parsedRemoteInfo.endsWith(': gone');
            const remoteBranchName = parsedRemoteInfo.replace(': gone', '');

            item.isMerged = isMerged;
            item.remoteName = remoteBranchName;
          }
        })
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        .filter((item) => (item as any).isMerged)
        .value();
    const branchNames = chain(localBranchesWithGoneRemotes).map('name').value();

    if (!branchNames.length) {
      console.log('PROJECT IS CLEAN! WELL DONE!');
      process.exit(0);
      return;
    }

    // interaction!
    const table = new Table({
      head: ['Branch Name', 'Origin Branch Name', 'Origin Branch Status'],
      chars
    });

    localBranchesWithGoneRemotes
      .forEach((item) => table.push([item.name, item.remoteName, 'GONE']));

    console.log(`${table.toString()}\n`);

    const ACTION_ANSWERS = {
      PRUNE_ALL: 'prune all gone branches',
      SELECT_BRANCHES: 'Selected Individual branches to delete'
    };

    const answers = await inquirer
      .prompt([
        {
          type: 'list',
          name: 'whatToDo',
          message: [
            'These branches have been deleted from the origin. ',
            'What do you want to do with them?'
          ].join(''),
          choices: chain(ACTION_ANSWERS).values().value()
        }
      ]);

    if (answers.whatToDo === ACTION_ANSWERS.PRUNE_ALL) {
      await moveToAnotherBranchIfNeeded(branchSummaryResult, branchNames);

      const result = await git.deleteLocalBranches(branchNames, true);
      console.log('DONE');
      chain(result.all)
        .map((item) => getStatusString(item))
        .forEach((item) => console.log(item))
        .value();
      return;
    }

    if (answers.whatToDo === ACTION_ANSWERS.SELECT_BRANCHES) {
      const answers = await inquirer
        .prompt([
          {
            type: 'checkbox',
            message: [
              'These branches have been deleted from the origin. ',
              'Do you want to prune them?'
            ].join(''),
            name: 'pruneBranches',
            choices: branchNames
          }
        ]);

      await moveToAnotherBranchIfNeeded(branchSummaryResult, branchNames);
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const deleteBranchPromises: Promise<any>[] = answers.pruneBranches
        .map((branchName: string) => git.deleteLocalBranch(branchName, true));
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const result: any = await Promise.all(deleteBranchPromises);
      console.log('DONE');
      console.log(result);
      chain(result.all)
        .map((item) => getStatusString(item))
        .forEach((item) => console.log(item));
      return;
    }
  } catch (err) {
    console.error(err);
    process.exit(1);
    return;
  }
}

async function moveToAnotherBranchIfNeeded(
  branchSummaryResult: BranchSummary,
  branchesToDelete: string[]
) {
  const suspectedMainBranch = branchSummaryResult.all
    .find((branchName: string) => MAIN_BRANCHES.includes(branchName)) as string;
  const currentCheckedoutBranch = branchSummaryResult.current;

  // console.log('main branch:', suspectedMainBranch);

  if (branchesToDelete.includes(currentCheckedoutBranch)) {
    console.warn([
      `trying to delete checkedout branch ${ currentCheckedoutBranch }. `,
      `moving to ${ suspectedMainBranch }`
    ].join(''));
    await git.checkout(suspectedMainBranch);
  }
}

function getStatusString(item: { branch: string; success: boolean }) {
  return `${ item.branch }: ${ item.success ? 'DELETED' : 'FAILED' }`;
}
