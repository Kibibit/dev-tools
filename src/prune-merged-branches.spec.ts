/* eslint-disable @typescript-eslint/no-explicit-any */
import mockConsole from 'jest-mock-console';
import { mockProcessExit } from 'jest-mock-process';
import simpleGit from 'simple-git/promise';
import strip from 'strip-ansi';

import { pruneMergedBranches } from './prune-merged-branches';

describe('Prune Merged Branches', () => {
  let mockExit;
  let restoreConsole;

  beforeEach(() => {
    jest.clearAllMocks();
    mockExit = mockProcessExit();
    restoreConsole = mockConsole();
    (simpleGit as any).clearMocks();
  });

  afterEach(() => {
    mockExit.mockRestore();
    restoreConsole();
  });

  it('should print branches that are gone', async () => {
    await pruneMergedBranches();
    expect(console.log).toHaveBeenCalledTimes(1);
    expect(strip((console.log as jest.Mock).mock.calls[0][0]))
      .toMatchSnapshot();
  });

  it('should do nothing if no merged branches found', async () => {
    (simpleGit as any).branchReturnObj = {
      all: [ 'main' ],
      branches: {
        main: {
          current: false,
          name: 'main',
          commit: 'cd2ec48',
          label: '[origin/main: behind 16] ci(github): add pipelines'
        }
      },
      current: 'main',
      detached: false
    };
    await pruneMergedBranches();
    expect(console.log).toHaveBeenCalledTimes(1);
    expect(strip((console.log as jest.Mock).mock.calls[0][0]))
      .toMatchSnapshot();
  });

  it('should move to another branch if needed', async () => {
    // restoreConsole();
    (simpleGit as any).branchReturnObj = {
      all: [ 'main' ],
      branches: {
        main: {
          current: false,
          name: 'main',
          commit: 'cd2ec48',
          label: '[origin/main: behind 16] ci(github): add pipelines'
        },
        'release/fix-release': {
          current: false,
          name: 'release/fix-release',
          commit: '8f64df7',
          label: '[origin/release/fix-release: gone] blah'
        }
      },
      current: 'release/fix-release',
      detached: false
    };
    await pruneMergedBranches();
    expect(console.log).toHaveBeenCalledTimes(1);
    expect(strip((console.log as jest.Mock).mock.calls[0][0]))
      .toMatchSnapshot();
    expect(console.warn).toHaveBeenCalledTimes(1);
    expect(strip((console.warn as jest.Mock).mock.calls[0][0]))
      .toMatchSnapshot();
  });
});
