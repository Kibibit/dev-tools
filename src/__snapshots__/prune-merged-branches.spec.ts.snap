// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prune Merged Branches should do nothing if no merged branches found 1`] = `"PROJECT IS CLEAN! WELL DONE!"`;

exports[`Prune Merged Branches should move to another branch if needed 1`] = `
"╔═════════════════════╤════════════════════════════╤══════════════════════╗
║ Branch Name         │ Origin Branch Name         │ Origin Branch Status ║
╟─────────────────────┼────────────────────────────┼──────────────────────╢
║ release/fix-release │ origin/release/fix-release │ GONE                 ║
╚═════════════════════╧════════════════════════════╧══════════════════════╝
"
`;

exports[`Prune Merged Branches should move to another branch if needed 2`] = `"trying to delete checkedout branch release/fix-release. moving to main"`;

exports[`Prune Merged Branches should print branches failed to delete 1`] = `"release/fix-release: DELETED"`;

exports[`Prune Merged Branches should print branches failed to delete 2`] = `"main: FAILED"`;

exports[`Prune Merged Branches should print branches that are gone 1`] = `
"╔══════════════════════════╤═════════════════════════════════╤══════════════════════╗
║ Branch Name              │ Origin Branch Name              │ Origin Branch Status ║
╟──────────────────────────┼─────────────────────────────────┼──────────────────────╢
║ release/fix-release      │ origin/release/fix-release      │ GONE                 ║
╟──────────────────────────┼─────────────────────────────────┼──────────────────────╢
║ release/new-main-release │ origin/release/new-main-release │ GONE                 ║
╚══════════════════════════╧═════════════════════════════════╧══════════════════════╝
"
`;

exports[`Prune Merged Branches should print branches that are gone 2`] = `"DONE"`;

exports[`Prune Merged Branches should print branches that are gone 3`] = `"release/fix-release: DELETED"`;
